File: java/src/main/java/io/pingui/monitor/AlertRateLimiter.java
package io.pingui.monitor;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Per-host rolling one-hour alert rate limit (P10-040 / ADR_ALERTS). */
public final class AlertRateLimiter {
    private static final double WINDOW_SECONDS = 3600.0;

    private final int maxPerHour;
    private final Map<String, List<Double>> history = new HashMap<>();

    public AlertRateLimiter(int maxPerHour) {
        if (maxPerHour < 1) {
            throw new IllegalArgumentException("maxPerHour must be >= 1");
        }
        this.maxPerHour = maxPerHour;
    }

    public boolean allow(String host) {
        return allow(host, System.currentTimeMillis() / 1000.0);
    }

    boolean allow(String host, double nowSeconds) {
        double windowStart = nowSeconds - WINDOW_SECONDS;
        List<Double> entries = new ArrayList<>();
        for (double timestamp : history.getOrDefault(host, List.of())) {
            if (timestamp > windowStart) {
                entries.add(timestamp);
            }
        }
        if (entries.size() >= maxPerHour) {
            history.put(host, entries);
            return false;
        }
        entries.add(nowSeconds);
        history.put(host, entries);
        return true;
    }
}

File: src/pingui/monitor/alert_rate_limiter.py
"""Per-host alert rate limiting."""

from __future__ import annotations

import time


class AlertRateLimiter:
    """Allow at most ``max_per_hour`` alerts per host in a rolling hour window."""

    def __init__(self, max_per_hour: int = 10) -> None:
        if max_per_hour < 1:
            msg = "max_per_hour must be >= 1"
            raise ValueError(msg)
        self._max_per_hour = max_per_hour
        self._history: dict[str, list[float]] = {}

    def allow(self, host: str, *, now: float | None = None) -> bool:
        """Return True when an alert for ``host`` may be sent."""
        ts = time.time() if now is None else now
        window_start = ts - 3600.0
        entries = [t for t in self._history.get(host, ()) if t > window_start]
        if len(entries) >= self._max_per_hour:
            self._history[host] = entries
            return False
        entries.append(ts)
        self._history[host] = entries
        return True
